### PR TITLE
Remove vm_id and zone requirement to start agent.

### PR DESCRIPTION
**Pull request by bmoyles0117**

### CODE REVIEW

File: lib/fluent/plugin/out_google_cloud.rb [hunk 1]
```diff
@@ -1013,8 +1015,10 @@ def set_required_metadata_variables
       # All metadata parameters must now be set.
       missing = []
       missing << 'project_id' unless @project_id
-      missing << 'zone' unless @zone
-      missing << 'vm_id' unless @vm_id
+      if @platform != Platform::OTHER
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply lib/fluent/plugin/out_google_cloud.rb]: I think rubocop only complains when you do "if not", I found a way to sneak around it :)

File: test/plugin/base_test.rb [hunk 2]
```diff
@@ -84,23 +84,23 @@ def test_configure_custom_metadata
...
-      CONFIG_MISSING_METADATA_ZONE => ['zone'],
-      CONFIG_MISSING_METADATA_VM_ID => ['vm_id'],
-      CONFIG_MISSING_METADATA_ALL => %w(project_id zone vm_id)
-    }.each_with_index do |(config, parts), index|
-      exception_count = 0
+    [[CONFIG_MISSING_METADATA_PROJECT_ID, ['project_id'], false],
+     [CONFIG_MISSING_METADATA_ZONE, [], true],
+     [CONFIG_MISSING_METADATA_VM_ID, [], true],
+     [CONFIG_MISSING_METADATA_ALL, ['project_id'], false]
+    ].each_with_index do |(config, parts, expected_valid_configuration), index|
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply test/plugin/base_test.rb]: Updated

File: test/plugin/base_test.rb [hunk 3]
```diff
@@ -84,23 +84,23 @@ def test_configure_custom_metadata
   def test_configure_invalid_metadata_missing_parts
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply test/plugin/base_test.rb]: Updated

File: lib/fluent/plugin/out_google_cloud.rb [hunk 4]
```diff
@@ -539,10 +539,12 @@ def configure(conf)
         @write_request = method(:write_request_via_rest)
       end
 
-      # Log an informational message containing the Logs viewer URL
-      @log.info 'Logs viewer address: https://console.cloud.google.com/logs/',
-                "viewer?project=#{@project_id}&resource=#{@resource.type}/",
-                "instance_id/#{@vm_id}"
+      if @vm_id
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply lib/fluent/plugin/out_google_cloud.rb]: Updated

File: lib/fluent/plugin/out_google_cloud.rb [hunk 5]
```diff
@@ -1013,8 +1015,10 @@ def set_required_metadata_variables
       # All metadata parameters must now be set.
       missing = []
       missing << 'project_id' unless @project_id
-      missing << 'zone' unless @zone
-      missing << 'vm_id' unless @vm_id
+      if @platform != Platform::OTHER
+        missing << 'zone' unless @zone
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Doesn't this go away altogether in #241?

Comment (bmoyles0117) [reply lib/fluent/plugin/out_google_cloud.rb]: Yes, but for now Ling has asked that we keep it, and I agree with this for doing the minimally disruptive action until we're ready to follow through with a larger refactoring.

File: test/plugin/base_test.rb [hunk 6]
```diff
@@ -81,26 +81,26 @@ def test_configure_custom_metadata
...
-    }.each_with_index do |(config, parts), index|
-      exception_count = 0
+    [[CONFIG_MISSING_METADATA_PROJECT_ID, ['project_id'], false],
+     [CONFIG_MISSING_METADATA_ZONE, [], true],
+     [CONFIG_MISSING_METADATA_VM_ID, [], true],
+     [CONFIG_MISSING_METADATA_ALL, ['project_id'], false]
+    ].each_with_index do |(config, missing_parts, is_valid_config), index|
       begin
         create_driver(config)
+
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: This blank line is confusing.

File: test/plugin/base_test.rb [hunk 7]
```diff
@@ -81,26 +81,26 @@ def test_configure_custom_metadata
...
+        assert_true is_valid_config, "Index #{index} failed."
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: The message should probably be more descriptive, e.g.: `"Invalid config at index #{index} should have raised an error."`. Same below, e.g.: `"Valid config at index #{index} should not have raised an error #{error}."`.

File: test/plugin/base_test.rb [hunk 8]
```diff
@@ -81,26 +81,26 @@ def test_configure_custom_metadata
...
       rescue Fluent::ConfigError => error
+        assert_false is_valid_config, "Index #{index} failed."
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: Add the error info (I've mentioned this above, too).

File: lib/fluent/plugin/out_google_cloud.rb [hunk 9]
```diff
@@ -539,10 +539,12 @@ def configure(conf)
         @write_request = method(:write_request_via_rest)
       end
 
-      # Log an informational message containing the Logs viewer URL
-      @log.info 'Logs viewer address: https://console.cloud.google.com/logs/',
-                "viewer?project=#{@project_id}&resource=#{@resource.type}/",
-                "instance_id/#{@vm_id}"
+      if @platform != Platform::OTHER
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply lib/fluent/plugin/out_google_cloud.rb]: I believe this is logically equivalent, and I'd err on the side of ensuring we only avoid validation of OTHER.

[the remark this answers is not in the document]
Comment (igorpeshansky) [reply lib/fluent/plugin/out_google_cloud.rb]: This wasn't about validation, this is about printing the Logs Viewer URL...

**Comment by qingling128** [comment]
What is the status of this PR? I thought it's ready to merge at some point?

**Comment by igorpeshansky** [comment]
I think this was rebased and needs re-approval...

**Comment by qingling128** [comment]
Is this branch ready to be deleted?

**Pull request merged**